### PR TITLE
[Container Registry] Acr delete artifact

### PR DESCRIPTION
**Pull request by seankane-msft**

### CODE REVIEW

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py [hunk 1]
```diff
@@ -49,10 +60,13 @@ def __init__(self, **kwargs):
         self.created_on = kwargs.get("created_on", None)
         self.digest = kwargs.get("digest", None)
         self.last_updated_on = kwargs.get("last_updated_on", None)
-        self.manifest_properties = kwargs.get("manifest_properties", None)
         self.operating_system = kwargs.get("operating_system", None)
+        self.references = kwargs.get("references", None)
         self.size = kwargs.get("size", None)
         self.tags = kwargs.get("tags", None)
+        self.content_permissions = kwargs.get("content_permissions", None)
```
Comment (swathipil) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py:67]: I believe `get` already defaults to `None` if key is not in there, so second arg `None` is not needed.
(not saying this needs to be changed or anything)

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_repository_client.py [hunk 2]
```diff
@@ -114,35 +112,36 @@ def get_tag_properties(self, tag, **kwargs):
...
-        #     n=kwargs.get("n", None),
-        #     orderby=kwargs.get("orderby"),
-        # )  # ,
-        # # cls=lambda objs: [RegistryArtifacts._from_generated(x) for x in objs])
-
-        # return RegistryArtifactProperties._from_generated(artifacts)
+        # GET /acr/v1/{name}/_manifests
+        last = kwargs.pop("last", None)
+        n = kwargs.pop("page_size", None)
+        orderby = kwargs.pop("order_by", None)
```
Comment (swathipil) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_repository_client.py:135]: I believe default return if key not provided is `None` so you don't need to provide that second arg.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_repository_client.py [hunk 3]
```diff
@@ -114,35 +112,36 @@ def get_tag_properties(self, tag, **kwargs):
...
+            self.repository,
+            last=last,
+            n=n,
+            orderby=orderby,
+            cls=lambda objs: [
+                RegistryArtifactProperties._from_generated(x) for x in objs  # pylint: disable=protected-access
+            ],
+        )
 
     def list_tags(self, **kwargs):
```
Comment (swathipil) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_repository_client.py:146]: `param`s are specified in the docstring, but it looks like they should be `keyword`?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py [hunk 4]
```diff
@@ -6,27 +6,38 @@
...
 
 if TYPE_CHECKING:
-    from ._generated.models import ManifestChangeableAttributes, ManifestAttributesBase
+    from ._generated.models import ManifestAttributesBase
     from ._generated.models import RepositoryProperties as GeneratedRepositoryProperties
+    from ._generated.models import TagProperties as GeneratedTagProperties
 
 
 class ContentPermissions(object):
     def __init__(self, **kwargs):
```
Comment (swathipil) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py:18]: if part of public API, docstring with description + keywords? (but maybe you know this already or are working on it.)

**Review by swathipil — COMMENTED** [review]
left some comments about things you probably already know

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_repository_client.py [hunk 5]
```diff
@@ -44,45 +50,40 @@ def delete(self, **kwargs):
...
 
-    def delete_registry_artifact(self, digest):
+    def delete_registry_artifact(self, digest, **kwargs):
         # type: (str) -> None
-        """Delete a registry artifact
+        """Delete a registry artifact. A registry artifact can only be deleted from the digest
 
         :param digest: The digest of the artifact to be deleted
         :type digest: str
         :returns: None
```
[the remark this answers is not in the document]
Comment (seankane-msft) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_repository_client.py:59]: I need to update docstrings, but I will probably do this in a separate PR.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_exchange_client.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        pass
+
+
+class ACRExchangeClient(object):
+    """Class for handling oauth authentication requests
+
+    :param endpoint: Azure Container Registry endpoint
+    :type endpoint: str
+    :param credential: AAD Token for authenticating requests with Azure
+    :type credential: :class:`azure.identity.DefaultTokenCredential`
```
[the remark this answers is not in the document]
Comment (seankane-msft) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_exchange_client.py:39]: This client is internal only but yes the docstrings need improvement.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py [hunk 7]
```diff
@@ -127,7 +127,7 @@ async def list_registry_artifacts(self, **kwargs) -> AsyncItemPaged[RegistryArti
         last = kwargs.pop("last", None)
         n = kwargs.pop("page_size", None)
         orderby = kwargs.pop("order_by", None)
-        return await self._client.container_registry_repository.get_manifests(
+        return self._client.container_registry_repository.get_manifests(
```
Comment (iscai-msft) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py:130]: why is the `await` gone

File: sdk/containerregistry/azure-containerregistry/tests/test_container_registry_client.py [hunk 8]
```diff
@@ -19,12 +19,14 @@
 from azure.identity import DefaultAzureCredential
 
 from testcase import ContainerRegistryTestClass
+from constants import TO_BE_DELETED
```
Comment (iscai-msft) [inline sdk/containerregistry/azure-containerregistry/tests/test_container_registry_client.py:22]: is this constant somethign you want exposed to users? I'm assuming not (which is how it currently is), but just double checking

**Review by iscai-msft — APPROVED** [review]

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py (hunk 7 above)
Comment (seankane-msft) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py:130]: The list methods don't return a coroutine, they return an AsyncItemPaged

File: sdk/containerregistry/azure-containerregistry/tests/test_container_registry_client.py (hunk 8 above)
Comment (seankane-msft) [reply sdk/containerregistry/azure-containerregistry/tests/test_container_registry_client.py:22]: No it is not, it's just so I have a consistent registry imported. I kept misspelling it and eventually made it a constant.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py (hunk 7 above)
Comment (iscai-msft) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py:130]: oh yeah lmao, ignore me

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py [hunk 9]
```diff
@@ -76,11 +76,11 @@ def _from_generated(cls, generated):
             created_on=generated.created_on,
             digest=generated.digest,
             last_updated_on=generated.last_updated_on,
-            manifest_properties=generated.manifest_properties,
             operating_system=generated.operating_system,
-            registry_artifacts=generated.registry_artifacts,
+            references=generated.references,
```
Comment (kristapratico) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py:80]: Just checking - should this be? 

```suggestion
            references=generated.registry_artifacts,
```

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py (hunk 7 above)
Comment (kristapratico) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py:130]: We can remove the "async" before the def list_registry_artifacts too

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py [hunk 10]
```diff
@@ -53,25 +53,25 @@ async def delete(self, **kwargs) -> None:
...
         :param digest: The digest of the artifact to be deleted
         :type digest: str
         :returns: None
         :raises: :class:~azure.core.exceptions.ResourceNotFoundError
         """
-        raise NotImplementedError("Has not been implemented")
+        await self._client.container_registry_repository.delete_manifest(self.repository, digest, **kwargs)
 
-    def delete_tag(self, tag: str, **kwargs) -> None:
+    async def delete_tag(self, tag: str, **kwargs) -> None:
```
Comment (kristapratico) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py:66]: haven't looked through all your PRs yet so might have missed it, but do you plan to add the distributed tracing decorators to your client methods?

Comment (seankane-msft) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_repository_client.py:70]: Added!

**Pull request merged**